### PR TITLE
[12.0][MIG] product_to_bizerba_scale

### PR DESCRIPTION
**Pull request by vvrossem**
### [Task](https://gestion.coopiteasy.be/web#id=3156&view_type=form&model=project.task&action=479)

This module
1) merges the following modules into one: 
- [product_to_scale_bizerba](https://github.com/coopiteasy/addons/tree/9.0/product_to_scale_bizerba)
- [product_to_scale_bizerba_extended](https://github.com/coopiteasy/addons/tree/9.0/product_to_scale_bizerba_extended)
2) migrates them to 12.0

Please review from commit ~~ebaf64be043755a33b1b909916173c4478e0a912~~ 7bb1e700d99d796cc0a3056ea565c961901615d3. My apologies to the reviewers : for better reviewing, I should have done two separate commits for
1) merging the two modules into one
2) migration to 10.0

### CODE REVIEW

File: product_to_bizerba_scale/views/product_template_view.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+<odoo>
+    <record id="product_template_form_view" model="ir.ui.view">
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="product.product_template_form_view"/>
+        <field name="arch" type="xml">
+            <notebook position="inside">
+                <page name="scale" string="Scale">
+                    <group name="scale_communication" string="Scale Communication">
+                        <group>
+                            <!-- <field name="scale_group_id" attrs="{'required':[('to_weight','=',True)]}"/> -->
```
Comment (houssine78) [inline product_to_bizerba_scale/views/product_template_view.xml]: why commenting or it was already commented ?

**Review by houssine78 — APPROVED** [review]

File: product_to_bizerba_scale/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# Copyright (C) 2016-Today GRAP (http://www.grap.coop)
+# © 2017 Coop IT Easy (http://www.coopiteasy.be)
+# @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
+# @author: Houssine BAKKALI (https://github.com/houssine78)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
Comment (houssine78) [inline product_to_bizerba_scale/__manifest__.py:5]: Could you add coop it easy copy right on all the py and limite the one from la louve to 2018.
Also make it comply with the CIE coopyright header.
thanks !

File: product_to_bizerba_scale/data/ir_cron.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<!--
+    Copyright (C) 2016-Today: La Louve (<http://www.lalouve.net/>)
```
Comment (houssine78) [inline product_to_bizerba_scale/data/ir_cron.xml]: not today 2018

File: product_to_bizerba_scale/data/ir_cron.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
```
Comment (houssine78) [inline product_to_bizerba_scale/data/ir_cron.xml]: i think it's better to remove the copyright from the xml files

File: product_to_bizerba_scale/demo/product_scale_group.xml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+<odoo>
+
+    <record id="demo_scale_group_fruits" model="product.scale.group">
+        <field name="name">Fruits</field>
+        <field name="scale_system_id" ref="demo_scale_system" />
+        <field name="external_identity">1</field>
+    </record>
+
+    <record id="demo_scale_group_vegetables" model="product.scale.group">
+        <field name="name">Végetables</field>
```
Comment (houssine78) [inline product_to_bizerba_scale/demo/product_scale_group.xml]: English doesn't take accent ;)

File: product_to_bizerba_scale/models/product_scale_log.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+    sent = fields.Boolean(
+        string='Is Sent'
+    )
+    last_send_date = fields.Datetime(
+        string='Last Send Date'
+    )
+
+    @api.noguess
+    def _auto_init(self):
+        # FIXME on install -> psycopg2.ProgrammingError: relation "product_scale_log" does not exist
```
Comment (houssine78) [inline product_to_bizerba_scale/models/product_scale_log.py:80]: didn't found a fix for this ?

File: product_to_bizerba_scale/views/product_template_view.xml (hunk 1 above)
Comment (vvrossem) [reply product_to_bizerba_scale/views/product_template_view.xml]: it was already commented in [product_to_scale_bizerba_extended (9.0)](https://github.com/coopiteasy/addons/blame/9.0/product_to_scale_bizerba_extended/views/product_view.xml#L12)

File: product_to_bizerba_scale/__manifest__.py (hunk 2 above)
Comment (vvrossem) [reply product_to_bizerba_scale/__manifest__.py:5]: done :)

File: product_to_bizerba_scale/data/ir_cron.xml (hunk 3 above)
Comment (vvrossem) [reply product_to_bizerba_scale/data/ir_cron.xml]: done :)

File: product_to_bizerba_scale/data/ir_cron.xml (hunk 4 above)
Comment (vvrossem) [reply product_to_bizerba_scale/data/ir_cron.xml]: done :)

File: product_to_bizerba_scale/demo/product_scale_group.xml (hunk 5 above)
Comment (vvrossem) [reply product_to_bizerba_scale/demo/product_scale_group.xml]: :baguette_bread:

File: product_to_bizerba_scale/models/product_scale_log.py (hunk 6 above)
Comment (vvrossem) [reply product_to_bizerba_scale/models/product_scale_log.py:80]: ouh ! I completely forgot about this one. Thank you for the reminder! I'll try to find a fix and ping you once it's done

Comment (vvrossem) [reply product_to_bizerba_scale/models/product_scale_log.py:80]: I've looked into it.
- `DELETE FROM product_scale_log` is an `Unsafe query: 'Delete' statement without 'where' clears all data in the table`
- we can use instead `DROP TABLE IF EXISTS product_scale_log`: it doesn't raise `psycopg2.ProgrammingError`

But, do we want this behavior ? If we execute an `update-module product_to_bizerba_scale`, all logs will be deleted, including the ones that might not have been sent to the FTP

**Comment by vvrossem** [comment]
@robinkeunen since you are familiar with csv and sftp, could you have a look on this PR ? :)

File: product_to_bizerba_scale/models/product_scale_log.py (hunk 6 above)
Comment (remytms) [reply product_to_bizerba_scale/models/product_scale_log.py:80]: I remember, that there is a problem with the `product.scale.log` table when performing an update on this module. So it blocks an update all.

I'm curious to see if this issue is still present after your refactoring. Maybe we should give it a try. Then bring back this hack if the problem still persist.

File: product_to_bizerba_scale/models/product_scale_system.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        required=True,
+        default='external_text.csv',
+        help="Pattern of the External Text file. Use % to include dated information. Ref: https://docs.python.org/2/library/time.html#time.strftime"
+    )
+    product_line_ids = fields.One2many(
+        comodel_name='product.scale.system.product.line',
+        inverse_name='scale_system_id',
+        string='Product Lines'
+    )
+    # TODO
```
Comment (remytms) [inline product_to_bizerba_scale/models/product_scale_system.py]: TODO what ?

File: product_to_bizerba_scale/models/product_scale_system_product_line.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        selection=_TYPE_SELECTION,
+        string='Type'
+    )
+    field_id = fields.Many2one(
+        comodel_name='ir.model.fields',
+        string='Product Field',
+        domain="[('model', 'in', ['product.template'])]"
+    )
+
+    # TODO Improve. Set domain, depending on the other field
```
Comment (remytms) [inline product_to_bizerba_scale/models/product_scale_system_product_line.py:47]: I suppose that this TODO comes from the 9.0 version.

File: product_to_bizerba_scale/models/product_scale_system_product_line.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        default=0
+    )
+    multiline_separator = fields.Char(
+        string='Separator for Multiline',
+        help="Used if type is 'Text Field' or 'External Text Constant', to indicate wich text will be used to mention "
+             "break lines.",
+        default='\n'
+    )
+
+    # TODO Improve. Set contrains.
```
Comment (remytms) [inline product_to_bizerba_scale/models/product_scale_system_product_line.py:77]: Same here.

File: product_to_bizerba_scale/models/product_template.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+# Copyright 2017-Today Coop IT Easy SCRLfs
+#   - Houssine BAKKALI <houssine@coopiteasy.be>
+#   - Rémy TAYMANS <remy@coopiteasy.be>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models, _
+from odoo.addons import decimal_precision as dp
+
+# TODO: scale_category is defined in beesdoo_product but this module do
+#       not depend on it. Find a way to configure these fields.
```
Comment (remytms) [inline product_to_bizerba_scale/models/product_template.py:12]: Ok, for now,  add `beesdoo_product` in the dependencies and leave this comment here. :)

File: product_to_bizerba_scale/models/product_template.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    @api.model
+    def create(self, vals):
+        product = super(ProductTemplate, self).create(vals)
+        if product.is_in_scale():
+            product._send_to_scale_bizerba('create')
+        return product
+
+    @api.multi
+    def write(self, vals):
+        defered = {}
```
Comment (remytms) [inline product_to_bizerba_scale/models/product_template.py]: I have made a type for `defered`, it should be spelled `deferred`. I will be nice if you can fix that. :)

File: product_to_bizerba_scale/views/product_scale_log_view.xml [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+                <field name="product_id"/>
+                <field name="action"/>
+            </tree>
+        </field>
+    </record>
+
+    <record id="view_product_scale_log_form" model="ir.ui.view">
+        <field name="model">product.scale.log</field>
+        <field name="arch" type="xml">
+            <form name="scale_log" string="Scale Log" version="7.0">
```
Comment (remytms) [inline product_to_bizerba_scale/views/product_scale_log_view.xml]: Does the attribute `version` make sense now ?

File: product_to_bizerba_scale/views/product_scale_system_view.xml [hunk 13]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<odoo>
+
+    <record id="view_product_scale_system_form" model="ir.ui.view">
+        <field name="model">product.scale.system</field>
+        <field name="arch" type="xml">
+            <form name="scale_system" string="Scale System" version="7.0">
```
Comment (remytms) [inline product_to_bizerba_scale/views/product_scale_system_view.xml]: Same here.

File: product_to_bizerba_scale/views/product_template_view.xml (hunk 1 above)
Comment (remytms) [reply product_to_bizerba_scale/views/product_template_view.xml]: Remove it. ;)

**Review by remytms — APPROVED** [review]
Some comment to be perfect. But anyway its good for me. :)

**Comment by robinkeunen** [comment]
@vvrossem I used a different library (paramiko) so I don't have much to add :-)

File: product_to_bizerba_scale/models/product_scale_system.py (hunk 7 above)
Comment (vvrossem) [reply product_to_bizerba_scale/models/product_scale_system.py]: oops, forgot to remove it ! -> removed

File: product_to_bizerba_scale/models/product_scale_system_product_line.py (hunk 8 above)
Comment (vvrossem) [reply product_to_bizerba_scale/models/product_scale_system_product_line.py:47]: exactly :) I leave it

File: product_to_bizerba_scale/models/product_scale_system_product_line.py (hunk 9 above)
Comment (vvrossem) [reply product_to_bizerba_scale/models/product_scale_system_product_line.py:77]: same as above :)

File: product_to_bizerba_scale/models/product_template.py (hunk 10 above)
Comment (vvrossem) [reply product_to_bizerba_scale/models/product_template.py:12]: replaced `product` with `beesdoo_product`. Should we move this module to `beescoop/obeesdoo`?

File: product_to_bizerba_scale/models/product_template.py (hunk 11 above)
Comment (vvrossem) [reply product_to_bizerba_scale/models/product_template.py]: renamed

File: product_to_bizerba_scale/views/product_scale_log_view.xml (hunk 12 above)
Comment (vvrossem) [reply product_to_bizerba_scale/views/product_scale_log_view.xml]: removed

File: product_to_bizerba_scale/views/product_scale_system_view.xml (hunk 13 above)
Comment (vvrossem) [reply product_to_bizerba_scale/views/product_scale_system_view.xml]: removed

File: product_to_bizerba_scale/views/product_template_view.xml (hunk 1 above)
Comment (vvrossem) [reply product_to_bizerba_scale/views/product_template_view.xml]: done

**Pull request merged**